src/Storages/ExternalTable/ClickHouse/ClickHouseSource.cpp

```diff
@@ -0,0 +1,74 @@
+#include <Common/quoteString.h>
+#include <Storages/ExternalTable/ClickHouse/ClickHouseSource.h>
+
+namespace DB
+{
+
+namespace ExternalTable
+{
+
+namespace
+{
+String constructSelectQuery(const String & database, const String & table, const Block & header)
+{
+    assert(header.columns());
+    const auto & col_names = header.getNames();
+
+    auto query = "SELECT " + backQuoteIfNeed(col_names[0]);
+    for (const auto & name : std::vector<String>(std::next(col_names.begin()), col_names.end()))
+        query.append(", " + backQuoteIfNeed(name));
+    query.append(" FROM " + (database.empty() ? "" : backQuoteIfNeed(database) + ".") + table);
+
+    return query;
+}
+
+}
+
+ClickHouseSource::ClickHouseSource(
+    const String & database,
+    const String & table,
+    const Block & header,
+    std::unique_ptr<ClickHouseClient> client_,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    ContextPtr context_,
+    Poco::Logger * logger_)
+    : ISource(header, true, ProcessorID::ClickHouseSourceID)
+    , client(std::move(client_))
+    , query(constructSelectQuery(database, table, header))
+    , context(context_)
+    , logger(logger_)
+{
+}
+
+Chunk ClickHouseSource::generate()
+{
+    if (isCancelled())
+    {
+        if (started)
+            client->cancelQuery();
+
+        return {};
+    }
+
+    if (!started)
+    {
+        started = true;
+        client->executeQuery(query);
+    }
+
+    LOG_INFO(logger, "polling data");
```

Comment:
Let's comment this log line since it will be too busy

Reply:
Removed.